### ISSUE TITLE
Brain extraction pipeline from antsCorticalThickness to ANTsR?

### ISSUE THREAD

**Issue by dorianps** [stnava/ANTsR#104]
I have some cases where brain extraction with abpBranExtraction() does not work well. After looking better into it, looks like the registration to template does not work well, including pieces of skull. Since the skull has highest values, the next step of kmean segmentation thinks that is the only area with white matter, and things don't get better from there.

![image](https://cloud.githubusercontent.com/assets/9083517/16849278/b6dc31a6-49c7-11e6-86da-e607140defdc.png)

After trying several things, I ended up trying antsCorticalThickness.sh, which produced excellent brain extraction. 

![image](https://cloud.githubusercontent.com/assets/9083517/16849315/e356c9da-49c7-11e6-8b0a-6732dbfbbc8a.png)

I know the act pipeline may take a little longer but I think is worth it. How can I implement the same routine in ANTsR, what are the best steps? If we already have smth like that in ANTsR that would be better.

Dorian

**Comment by stnava** [comment]
probably just changing the registration from SyN to SyNCC ... but you'd
have to go line by line and compare to get a better idea.

possibly also comparing the N3 or N4 calls.


brian

**Comment by ntustison** [comment]
``antsCorticalThickness.sh`` uses ``antsBrainExtraction.sh`` so you can call the latter directly.  Alternatively, look at the ``-y`` option in ``antsCorticalThickness.sh`` to only run specific stages.

**Comment by dorianps** [comment]
`antsBrainExtraction.sh` produced the same erronous result as antsBrainExtraction(). The SyN to SyNCC change may make the difference. I am running a SyNCC now to see if that fixes things.

-y is a cool addition in `antsBrainExtraction`, my slightly old version of ANTs didn't have that.

Thank you.

**Comment by ntustison** [comment]
I'm guessing you ran it in adifferent way because antsCorticalThickness literally calls antsBrainExtraction.

**Comment by stnava** [comment]
@ntustison, @dorianps is using the antsr version which takes some shortcuts


brian

**Comment by dorianps** [comment]
We looked a bit into this with Phil. The differences are that:
1. I was calling `antsBrainExtraction.sh` without the -f option, while the call from within `antsCorticalThickness.sh` is with -f option. The mask we pass with -f removes most of the face on the template, which probably helps avoiding costs of face registration. I may need to add the -f option to the ANTsR function to make things work better.

2. We checked the initialization performed in R and looks ok, but the presence of the face might have taken a toll on the affine initalization, mostly by not rotating the back of the brain and the brainstem.

![image](https://cloud.githubusercontent.com/assets/9083517/16854446/e870a2e4-49de-11e6-8336-cf65612bc493.png)

![image](https://cloud.githubusercontent.com/assets/9083517/16854436/e1138d9a-49de-11e6-823b-a4c0834881d7.png)

3. Maybe of lesser importance, but just to keep track, there are two differences between the R and shell version. In R we use affineInitializer, this is commented out in the shell version, and now is used antsAI. Also the antsAI call uses Mattes since April instead of MI. Might be worth knowing this when we try to match the shell version. Currently there is no antsAI wrapper in ANTsR.

I will try `antsBrainExtraction.sh` things with the -f option, and if that works, will change `antsBrainExtrac tion()` in R to add the possibility of passing a skull-but-no-face mask. Is it ok to add that?

**Comment by dorianps** [comment]
Ok, simply adding the -f option to `antsBrainExtraction.sh` made the difference (at least for this subject). The mask now is exactly like that produced by `antsCorticalThickness.sh`. Remains only to adapt this to `antsBrainExtraction()` in R.

**Comment by stnava** [comment]
that should be easy - affineInitializer lets you pass a mask in ...


brian

**Comment by dorianps** [comment]
Adding the mask option and improving the registration resolved in good part the problems I had with some subjects.
I have created a pull request with the following fixes:
abpBrainExtraction can take now a registration mask (a fix request that was in the old code as well). The user can also select the type of registration to perform. SyN is same as before and default, but I have added SyNabp which matches the shell parameters.
SyNabp has been added as an option in antsRegistration.

https://github.com/stnava/ANTsR/pull/106
